### PR TITLE
add new webkit target

### PR DESCRIPTION
**Pull request by brettferdosi**
draft until the webkit repository is made public

### CODE REVIEW

**Comment by jrtc27** [comment]
Some of those hunks look like they should be pulled out and can be merged straight away?

File: pycheribuild/projects/cross/crosscompileproject.py [hunk 1]
```diff
@@ -50,7 +50,8 @@
 # avoids repeating this for every target than can be cross-built
 class CrossCompileMixin(object):
     do_not_add_to_targets = True
-    supported_architectures = CompilationTargets.ALL_SUPPORTED_CHERIBSD_AND_HOST_TARGETS
+    supported_architectures = (CompilationTargets.ALL_SUPPORTED_CHERIBSD_AND_HOST_TARGETS +
```
Comment (arichardson) [inline pycheribuild/projects/cross/crosscompileproject.py:53]: This should be a separate commit, since it adds all the -morello targets.

Comment (arichardson) [reply pycheribuild/projects/cross/crosscompileproject.py:53]: Also you should probably update ALL_SUPPORTED_CHERIBSD_AND_HOST_TARGETS instead.

File: pycheribuild/config/target_info.py [hunk 2]
```diff
@@ -275,6 +275,8 @@ def capability_size(self):
             return self.config.mips_cheri_bits // 8
         elif self.target.is_hybrid_or_purecap_cheri([CPUArchitecture.RISCV64]):
             return 16  # RISCV64 uses 128-bit capabilities
+        elif self.target.is_hybrid_or_purecap_cheri([CPUArchitecture.AARCH64]):
```
Comment (arichardson) [inline pycheribuild/config/target_info.py:278]: Can you split this into a separate commit?

File: pycheribuild/projects/cross/webkit.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
+#
```
Comment (arichardson) [inline pycheribuild/projects/cross/webkit.py:1]: This looks pretty much identical to the other version that I wrote, maybe add a 
Co-authored-by: Alex Richardson <Alexander.Richardson@cl.cam.ac.uk>` line to the commit message.

**Review by arichardson — COMMENTED** [review]
Generally LGTM, but we should probably split the non-webkit changes into separate commits.

File: pycheribuild/projects/cross/webkit.py (hunk 3 above)
Comment (jrtc27) [reply pycheribuild/projects/cross/webkit.py:1]: In that case can we make them share code?

**Comment by arichardson** [comment]
This will need a few smaller changes to get `./cheribuild.py -p __run_everything__` to pass. Alternatively you could install the pre-push commit hook to run the tests locally before you push:
`ln -s ../../githooks/pre-push.sh .git/hooks/pre-push`

File: pycheribuild/projects/cross/webkit.py (hunk 3 above)
Comment (arichardson) [reply pycheribuild/projects/cross/webkit.py:1]: Sorry, I don't mean the existing the qtwebkit implementation, I meant the externally-managed private version that existed prior to open-sourcing the new webkit repo.

File: pycheribuild/config/compilation_targets.py [hunk 4]
```diff
@@ -611,7 +611,7 @@ def essential_compiler_and_linker_flags(self) -> typing.List[str]:
 
     @property
     def must_link_statically(self):
-        return True  # dynamic linking is still experimental
+        return False
```
Comment (arichardson) [inline pycheribuild/config/compilation_targets.py:618]: This is already the default in the parent class, so you could just remove the override.

**Review by arichardson — APPROVED** [review]
1 minor comment otherwise LGTM.

File: pycheribuild/config/compilation_targets.py [hunk 5]
```diff
@@ -973,7 +977,8 @@ class CompilationTargets(BasicCompilationTargets):
                                            CHERIBSD_MIPS_PURECAP, CHERIBSD_MIPS_HYBRID, CHERIBSD_MIPS_NO_CHERI]
     ALL_CHERIBSD_NON_MORELLO_TARGETS = ALL_CHERIBSD_MIPS_AND_RISCV_TARGETS + [CHERIBSD_AARCH64, CHERIBSD_X86_64]
     ALL_CHERIBSD_MORELLO_TARGETS = [CHERIBSD_MORELLO_PURECAP, CHERIBSD_MORELLO_HYBRID]
-    ALL_SUPPORTED_CHERIBSD_AND_HOST_TARGETS = ALL_CHERIBSD_NON_MORELLO_TARGETS + [BasicCompilationTargets.NATIVE]
+    ALL_SUPPORTED_CHERIBSD_AND_HOST_TARGETS = (ALL_CHERIBSD_NON_MORELLO_TARGETS + ALL_CHERIBSD_MORELLO_TARGETS +
```
Comment (jrtc27) [inline pycheribuild/config/compilation_targets.py:980]: No need for the parentheses

File: pycheribuild/projects/cross/qt5.py [hunk 6]
```diff
@@ -433,6 +433,7 @@ def __init__(self, config):
             # self.configure_args.append("--disable-extras")  # can't add this to host build, it will fail otherwise
             # We have modified the ICU data Makefile so that ICU builds a big endian data archive
             self.configure_args.append("--with-data-packaging=archive")
+            self.COMMON_FLAGS.append("-fPIC")  # needed for morello hybrid webkit not to complain about relocations
```
Comment (jrtc27) [inline pycheribuild/projects/cross/qt5.py:436]: Probably wants to be -fPIE not -fPIC, no?

Comment (jrtc27) [reply pycheribuild/projects/cross/qt5.py:436]: (although they should fix their toolchain to support PDEs properly... and an XXX: and maybe only do this for aarch64?)

File: pycheribuild/projects/cross/webkit.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            ENABLE_SAMPLING_PROFILER=False,
+            CHERI_PURE_CAPABILITY=self.crosscompile_target.is_cheri_purecap(),
+            )
+
+        if self.crosscompile_target.is_cheri_purecap():
+            # TODO: we can get this from the pre-processor instead
+            self.add_cmake_options(CHERI_CAPABILITY_SIZE=self.target_info.capability_size_in_bits)
+            self.add_cmake_options(ENABLE_JSHEAP_CHERI_OFFSET_REFS=self.jsheapoffsets)
+
+        # To allow benchmark comparability, we use system malloc and the contiguous arena for
```
Comment (jrtc27) [inline pycheribuild/projects/cross/webkit.py:110]:
```suggestion
        # To allow benchmark comparability, we use system malloc and the continuous arena for
```

File: pycheribuild/projects/cross/qt5.py (hunk 6 above)
Comment (arichardson) [reply pycheribuild/projects/cross/qt5.py:436]: I'm not sure if the build system produces any, but the flag also applies to shared libraries, so probably -fPIC is safer.

**Comment by arichardson** [comment]
> draft until the webkit repository is made public

We can merge this before the repo is public, we also have other targets that need non-public resources (e.g. spec2006).
